packages/asset-swapper/src/utils/swap_quote_calculator.ts

```diff
@@ -173,324 +166,75 @@ export class SwapQuoteCalculator {
-    operation: MarketOperation,
-): SwapQuoteOrdersBreakdown {
-    const orderAmounts =
-        operation === MarketOperation.Buy
-            ? orders.map(o => o.fill.totalMakerAssetAmount)
-            : orders.map(o => o.fill.totalTakerAssetAmount);
-    const amountsBySource: SwapQuoteOrdersBreakdown = {};
-    orders.forEach((o, i) => {
-        const source = o.fill.source;
-        amountsBySource[source] = orderAmounts[i].plus(amountsBySource[source] || 0);
-    });
-    const totalAmount = BigNumber.sum(0, ...orderAmounts);
+function getSwapQuoteOrdersBreakdown(fillAmountBySource: { [source: string]: BigNumber }): SwapQuoteOrdersBreakdown {
+    const totalFillAmount = BigNumber.sum(...Object.values(fillAmountBySource));
     const breakdown: SwapQuoteOrdersBreakdown = {};
-    for (const [source, amount] of Object.entries(amountsBySource)) {
-        breakdown[source] = amount.div(totalAmount);
-    }
+    Object.entries(fillAmountBySource).forEach(([source, fillAmount]) => {
+        breakdown[source] = fillAmount.div(totalFillAmount);
+    });
     return breakdown;
 }
 
-function getTakerAssetAmountBreakDown(
-    order: SignedOrderWithFillableAmounts,
-    takerAssetAmountWithFees: BigNumber,
-): { feeTakerAssetAmount: BigNumber; takerAssetAmount: BigNumber } {
-    if (isOrderTakerFeePayableWithTakerAsset(order)) {
-        const adjustedTakerAssetAmount = order.takerAssetAmount.plus(order.takerFee);
-        const filledRatio = takerAssetAmountWithFees.div(adjustedTakerAssetAmount);
-        const takerAssetAmount = filledRatio.multipliedBy(order.takerAssetAmount).integerValue(BigNumber.ROUND_CEIL);
-        return {
-            takerAssetAmount,
-            feeTakerAssetAmount: takerAssetAmountWithFees.minus(takerAssetAmount),
-        };
-    } else if (isOrderTakerFeePayableWithMakerAsset(order)) {
-        if (takerAssetAmountWithFees.isZero()) {
-            return {
-                takerAssetAmount: constants.ZERO_AMOUNT,
-                feeTakerAssetAmount: constants.ZERO_AMOUNT,
-            };
-        }
-        const takerFeeAmount = orderCalculationUtils.getTakerFeeAmount(order, takerAssetAmountWithFees);
-        const makerAssetFillAmount = orderCalculationUtils.getMakerFillAmount(order, takerAssetAmountWithFees);
-        const takerAssetAmount = takerFeeAmount
-            .div(makerAssetFillAmount)
-            .multipliedBy(takerAssetAmountWithFees)
-            .integerValue(BigNumber.ROUND_UP);
-        return {
-            takerAssetAmount,
-            feeTakerAssetAmount: takerAssetAmountWithFees.minus(takerAssetAmount),
-        };
-    }
+function fillResultsToQuoteInfo(fr: QuoteFillResult): SwapQuoteInfo {
     return {
-        feeTakerAssetAmount: constants.ZERO_AMOUNT,
-        takerAssetAmount: takerAssetAmountWithFees,
+        makerAssetAmount: fr.totalMakerAssetAmount,
+        takerAssetAmount: fr.takerAssetAmount,
+        totalTakerAssetAmount: fr.totalTakerAssetAmount,
```

Comment:
A consequence of this new approach is that _both_ the maker (in buys) and taker (in sells) asset amounts can theoretically deviate from the input amount slightly when fees are involved. In practice they're probably almost always the same though. So if we want to reference a fixed buy or sell amount in the API, we should be using the `quote.makerAssetAmount` or `quote.takerAssetFillAmount`, not the best or worst case values. Just something to note.